hummingbot/connector/exchange/gate_io/gate_io_exchange.py

```diff
@@ -0,0 +1,923 @@
+            for response, tracked_order in zip(responses, tracked_orders):
+                client_order_id = tracked_order.client_order_id
+                if isinstance(response, GateIoAPIError):
+                    err = response.error_payload
+                    if err.get('label') == 'ORDER_NOT_FOUND':
+                        self._order_not_found_records[client_order_id] = \
+                            self._order_not_found_records.get(client_order_id, 0) + 1
+                        if self._order_not_found_records[client_order_id] < self.ORDER_NOT_EXIST_CONFIRMATION_COUNT:
+                            # Wait until the order not found error have repeated a few times before actually treating
+                            # it as failed. See: https://github.com/CoinAlpha/hummingbot/issues/601
+                            continue
+                        self.trigger_event(MarketEvent.OrderFailure,
+                                           MarketOrderFailureEvent(
+                                               self.current_timestamp, client_order_id, tracked_order.order_type))
+                        self.stop_tracking_order(client_order_id)
+                    else:
+                        continue
+                elif "id" not in response:
+                    self.logger().info(f"_update_order_status id not in resp: {response}")
+                    continue
+                else:
+                    self._process_order_message(response)
+
+    def _process_order_message(self, order_msg: Dict[str, Any]):
+        """
+        Updates in-flight order and triggers cancellation or failure event if needed.
+        :param order_msg: The order response from either REST or web socket API (they are of the same format)
+        Example Order:
+        {
+            "id": "52109248977",
+            "text": "3",
+            "create_time": "1622638707",
+            "update_time": "1622638807",
+            "currency_pair": "BTC_USDT",
+            "type": "limit",
+            "account": "spot",
+            "side": "buy",
+            "amount": "0.001",
+            "price": "1999.8",
+            "time_in_force": "gtc",
+            "left": "0.001",
+            "filled_total": "0",
+            "fee": "0",
+            "fee_currency": "BTC",
+            "point_fee": "0",
+            "gt_fee": "0",
+            "gt_discount": true,
+            "rebated_fee": "0",
+            "rebated_fee_currency": "BTC",
+            "create_time_ms": "1622638707326",
+            "update_time_ms": "1622638807635",
+            ... optional params
+            "status": "open",
+            "event": "finish"
+            "iceberg": "0",
+            "fill_price": "0",
+            "user": 5660412,
+        }
+        """
+
+        # Call Update balances on every message to catch order create, fill and cancel - websocket doesn't support this.
+        safe_ensure_future(self._update_balances())
```

Comment:
I don't think it's a good idea to call update_balances here as it might interfere with the internal balance accounting mechanism I told you about.

Reply:
@Nullably is update balances automatically called when switching real time off then?

Reply:
No, the available balances are calculated by accounting for current inflight orders.  
When you use `safe_ensure_future` you don't get the execution immediate, you just start the async API call process. And you want to be careful with API rate limit as well.

Reply:
I've moved the update balance call to the end of the function: https://github.com/CoinAlpha/hummingbot/pull/3484/files?file-filters%5B%5D=.py#diff-589585025595c725ebc393ef67f553c926f35a9409829787b86536bc890e808aR704

Surely `safe_ensure_future` is what I want to use though as I don't want to block order creation returning until the REST balance has updated do I?